### PR TITLE
Fix on-insert decompression after schema changes

### LINKED ISSUE #5577: [Bug]: On-insert decompression after schema changes may not work properly
### What type of bug is this?

Incorrect result

### What subsystems and features are affected?

Compression

### What happened?

This could lead to the same kind of incorrect result issues as #5553 ; from a slightly different angle.

* compressed table with at least 1 segmentby column `(time, some_col, segmentby_col)`
* unique: `time,segmentby_col` 
* main table schema is altered - `some_col` is dropped
* insertion of duplicate row is not detected

-- 

what happens:
* [build_scankey](https://github.com/timescale/timescaledb/blob/d32224a914877f8f9342cca63859ca1d76d9bddb/tsl/src/compression/compression.c#L1823) works on 3 schema levels:
  * the `slot` is at the main hypertable level
  * the `decompressor->out_rel`  is at the standard inherited hypertable table level
    * `scankeys` are acquired at this level
  * the `decompressor->in_rel`  is at the compressed level
* during the `scankey` build; the actual value is extracted from the `slot`; but by using the `out_rel` level indexes ; which may lead to problems..even segfault if the drops are propelry planned :D

### TimescaleDB version affected

main

### PostgreSQL version used

15.2

### What operating system did you use?

Debian12

### What installation method did you use?

Source

### What platform did you run on?

Not applicable

### Relevant log output and stack trace

```bash
select count(1) from readings where battery_temperature=0.2;
 count 
-------
     2
(1 row)

select * from readings;
             time             | battery_temperature 
------------------------------+---------------------
 Fri Nov 11 03:11:11 2022 PST |                    
 Fri Nov 11 11:11:11 2022 PST |                 0.2
 Fri Nov 11 11:11:11 2022 PST |                 0.2
(3 rows)

-- unique check failure during decompression
select decompress_chunk(show_chunks('readings'),true);
ERROR:  duplicate key value violates unique constraint "_hyper_1_3_chunk_readings_uniq_idx"
```

### How can we reproduce the bug?

```bash
drop table if exists readings;
CREATE TABLE readings(
    "time"  TIMESTAMPTZ NOT NULL,
    battery_status  TEXT,
    battery_temperature  DOUBLE PRECISION
);

INSERT INTO readings ("time") VALUES ('2022-11-11 11:11:11-00');

SELECT create_hypertable('readings', 'time', chunk_time_interval => interval '12 hour', migrate_data=>true);

ALTER TABLE readings SET (timescaledb.compress,timescaledb.compress_segmentby = 'battery_temperature');
SELECT compress_chunk(show_chunks('readings'));

ALTER TABLE readings DROP COLUMN battery_status;
INSERT INTO readings ("time", battery_temperature) VALUES ('2022-11-11 11:11:11', 0.2);
SELECT readings FROM readings;

create unique index readings_uniq_idx on readings("time",battery_temperature);

SELECT decompress_chunk(show_chunks('readings'),true);
SELECT compress_chunk(show_chunks('readings'),true);

INSERT INTO readings ("time", battery_temperature) VALUES
    ('2022-11-11 11:11:11',0.2) -- same record as inserted 
;

select count(1) from readings where battery_temperature=0.2;

select * from readings;
-- unique check failure during decompression
select decompress_chunk(show_chunks('readings'),true);
```

case producing segfault:
```
drop table if exists readings;
CREATE TABLE readings(
    "time"  TIMESTAMPTZ NOT NULL,
    battery_status  TEXT,
    candy integer,
    battery_status2  TEXT,
    battery_temperature  TEXT
);

SELECT create_hypertable('readings', 'time', chunk_time_interval => interval '12 hour', migrate_data=>true);

ALTER TABLE readings SET (timescaledb.compress,timescaledb.compress_segmentby = 'battery_temperature');

ALTER TABLE readings DROP COLUMN battery_status;
ALTER TABLE readings DROP COLUMN battery_status2;
INSERT INTO readings ("time", candy,battery_temperature) VALUES ('2022-11-11 11:11:11', 88,'0.2');

create unique index readings_uniq_idx on readings("time",battery_temperature);

SELECT compress_chunk(show_chunks('readings'),true);

-- segfault
INSERT INTO readings ("time", candy,battery_temperature) VALUES
    ('2022-11-11 11:11:11',33,0.1) -- same record as inserted 
;

### PR DESCRIPTION
**Pull request by kgyrtkirk**
On compressed tables 3 schema levels are active simultaneously;
in build_scankeys all of them appears - and the hypertable level slot
was access using sibling table indices.
Multiple alteration of the table schema could lead to segfaults as well.

Fixes #5577

depends on #5553

### CODE REVIEW

File: tsl/src/compression/compression.c [hunk 1]
```diff
@@ -1989,6 +1991,7 @@ decompress_batches_for_insert(ChunkInsertState *cis, Chunk *chunk, TupleTableSlo
 
 	int num_scankeys;
 	ScanKeyData *scankeys = build_scankeys(chunk->fd.hypertable_id,
+										   chunk->hypertable_relid,
```
Comment (antekresic) [inline tsl/src/compression/compression.c:1994]: How about we just push the chunk to build_scankeys?

**Review by mkindahl — COMMENTED** [review]
The commit comment is very hard to follow. It would be good if you could rewrite it so that it is clear what the commit does and why it is necessary.

File: tsl/test/sql/compression_errors2.sql
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply tsl/test/sql/compression_errors2.sql:1]: I don't think we should put a lot of tests into the same file....its harder to follow/maintain/debug

I can put it in the other...but dropping the table to create anotherone with the same name is at least confusing

**Review by mkindahl — APPROVED** [review]

**Review by antekresic — APPROVED** [review]
